# This is a project about healthy heart.

In this project we will make a great model to predict heart disease.
We will use this model in other dataset, the second data is for train and the first is for apply our model.

We are going to work with two datasets, in each of these datasets have a column called target that means the diagnosis, we will use it to do the model and later the model will be used in another dataset to compare the predicted diagnosis with the old diagnosis that we have in the data, the target = 0 means patient healthy and target = 1 means patient with heart disease.

In [160]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn import metrics
from sklearn import preprocessing
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score

In [161]:
df1 = pd.read_csv('C:/Users/Vitorioso1/Documents/Project-A/Cardio/heart_statlog_cleveland_hungary_final.csv')
df1.head(3)

age  sex  chest pain type  resting bp s  cholesterol  fasting blood sugar  \
0   40    1                2           140          289                    0   
1   49    0                3           160          180                    0   
2   37    1                2           130          283                    0   

   resting ecg  max heart rate  exercise angina  oldpeak  ST slope  target  
0            0             172                0      0.0         1       0  
1            0             156                0      1.0         2       1  
2            1              98                0      0.0         1       0

In [162]:
df2 = pd.read_csv('C:/Users/Vitorioso1/Documents/Project-A/Cardio/heart.csv')
df2.head(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0

In [163]:
print(len(df1), len(df2))

1190 1025


##### Now let's change the name of some columns on the second data.

In [164]:
df2.columns = ['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol', 'fasting blood sugar', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak', 'slope', 'ca', 'thal', 'target']

In [165]:
df1.head(1)

age  sex  chest pain type  resting bp s  cholesterol  fasting blood sugar  \
0   40    1                2           140          289                    0   

   resting ecg  max heart rate  exercise angina  oldpeak  ST slope  target  
0            0             172                0      0.0         1       0

In [166]:
df2.head(1)

age  sex  chest pain type  resting bp s  cholesterol  fasting blood sugar  \
0   52    1                0           125          212                    0   

   resting ecg  max heart rate  exercise angina  oldpeak  slope  ca  thal  \
0            1             168                0      1.0      2   2     3   

   target  
0       0

##### Now let's create a variable with data that we will train, but before we need to do a test with some algorithms of classification in this data.

In [348]:
X_data = df2[['age', 'sex', 'resting bp s', 'cholesterol', 'fasting blood sugar', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak']]
y_data = df2['target']

In [349]:
X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size=0.6)

In [350]:
cv = svm.SVC(kernel='linear')

In [351]:
rf = RandomForestClassifier(n_estimators=600, criterion='entropy')

In [352]:
dc = DecisionTreeClassifier(criterion='entropy', max_depth=5)

In [353]:
scoresC = cross_val_score(cv, X_data, y_data, scoring='accuracy')
print('The accuracy of Support Virtual Machine in this dataset is:{0:.1f}%'.format(np.mean(scoresC)*100))

The accuracy of Support Virtual Machine in this dataset is:78.2%


In [354]:
scoresR = cross_val_score(rf, X_data, y_data, scoring='accuracy')
print('The accuracy of Random Forest in this dataset is:{0:.1f}%'.format(np.mean(scoresR)*100))

The accuracy of Random Forest in this dataset is:98.4%


##### As we can see, the model random forest will be effective in our data with 98.4% of accuracy.

In [355]:
rf.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=600)

In [356]:
r_pred = rf.predict(X_test)

In [357]:
r_pred[:10]

array([1, 0, 1, 0, 1, 1, 1, 0, 0, 0], dtype=int64)

In [358]:
print('The accuracy of our predicted data is:{0:.1f}%'.format(metrics.accuracy_score(y_test, r_pred)*100))

The accuracy of our predicted data is:93.0%


##### Now have a predictor of heart dieases with more than 90% of accuracy.

##### Let's use this model an a real dataset.

In [361]:
r2_pred = rf.predict(df1[['age', 'sex', 'resting bp s', 'cholesterol', 'fasting blood sugar', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak']])

In [362]:
df1['diagnosis'] = pd.DataFrame(r2_pred)

In [393]:
df1.head()

age  sex  chest pain type  resting bp s  cholesterol  fasting blood sugar  \
0   40    1                2           140          289                    0   
1   49    0                3           160          180                    0   
2   37    1                2           130          283                    0   
3   48    0                4           138          214                    0   
4   54    1                3           150          195                    0   

   resting ecg  max heart rate  exercise angina  oldpeak  ST slope  target  \
0            0             172                0      0.0         1       0   
1            0             156                0      1.0         2       1   
2            1              98                0      0.0         1       0   
3            0             108                1      1.5         2       1   
4            0             122                0      0.0         1       0   

   diagnosis  
0          1  
1          1  
2          1  
3          0  
4          1

In [377]:
len(df1)

1190

##### Our model have more than 90% of accuracy, but the column predicted don't have the same value than the column target.

##### We can see how many times the predicted value and the actual value will be with the same value.

In [396]:
x = df1[['target', 'diagnosis']][df1[['target', 'diagnosis']].iloc[:]==1]

In [414]:
x.tail(3)

target  diagnosis
1187     1.0        NaN
1188     1.0        1.0
1189     NaN        1.0

In [399]:
xx = x.dropna()

In [413]:
xx.head(3)

target  diagnosis
1      1.0        1.0
30     1.0        1.0
41     1.0        1.0

In [406]:
df1[['target', 'diagnosis']].iloc[30], df1[['target', 'diagnosis']].iloc[41]

(target       1
 diagnosis    1
 Name: 30, dtype: int64,
 target       1
 diagnosis    1
 Name: 41, dtype: int64)

In [415]:
y = df1[['target','diagnosis']][df1[['target', 'diagnosis']].iloc[:]==0]
yy = y.dropna()

In [416]:
yy.head(3)

target  diagnosis
21     0.0        0.0
26     0.0        0.0
39     0.0        0.0

In [417]:
len(yy)

86

In [418]:
len(xx)

111

In [419]:
len(df1)

1190

##### In this data we have 1190 lines, the predicted diagnosis agreed with the column target few times, that is 111 in patient with heart disease and 86 with healthy heart. 

### We have contradictory data, so what we should know is to compare the point at which the columns are agreeing is very important. Unfortunately we have only this contradictory data.

### taking the data seriously we can say that the doctor of the first data or the second data made big mistakes in the diagnosis.

 ...

# Thank you for work with me.